### ISSUE TITLE
Creating libaws-iot-sdk-cpp for Android

### ISSUE THREAD

**Issue by chaunnt** [aws/aws-iot-device-sdk-cpp#35]
Can I use this library for Android project ?
I have C++ project that will use this library.
Everything in Linux 14.04 desktop is OK
When I cross-compile this project for Android
it can not find -laws-iot-sdk-cpp
I also put library to lib folder of toolchain or /usr/local/lib and include all header file of library.
Questions are:
- Can I use this library for C++ project but cross-compile for Android ? (I'm using Qt)
- Can I build this library as shared library (file .so) with cross compiler ?

Note: 
1. I added some texts to cmake files but failed

> set(CMAKE_SYSTEM_NAME  Android)
> 
> set(CMAKE_SYSTEM_VERSION 1)
> 
> set(CMAKE_C_COMPILER /opt/arm-linux-androideabi-4.8/bin/arm-linux-androideabi-gcc)
> set(CMAKE_CXX_COMPILER /opt/arm-linux-androideabi-4.8/bin/arm-linux-androideabi-g++)

2. I also tried Cross-Compile with Cmake command (Removed text in cmake files of Note .1

> /opt/cmake/bin/cmake ../. -DCMAKE_TOOLCHAIN_FILE=//home/nntchau/Android/Sdk/ndk-bundle/build/cmake/android.toolchain.cmake

Output is as below

> -- Android: Selected GCC toolchain 'arm-linux-androideabi-4.9'
> -- Check for working CXX compiler: //home/nntchau/Android/Sdk/ndk-bundle/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/bin/arm-linux-androideabi-g++
> -- Check for working CXX compiler: //home/nntchau/Android/Sdk/ndk-bundle/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/bin/arm-linux-androideabi-g++ -- broken
> CMake Error at /opt/cmake/share/cmake-3.8/Modules/CMakeTestCXXCompiler.cmake:44 (message):
>   The C++ compiler
>   "//home/nntchau/Android/Sdk/ndk-bundle/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/bin/arm-linux-androideabi-g++"
>   is not able to compile a simple test program.
> 
>   It fails with the following output:
> 
>    Change Dir: /home/nntchau/aws-iot-device-sdk-cpp-1.1.1/build_share/CMakeFiles/CMakeTmp
> 
>   
> 
>   Run Build Command:"/usr/bin/make" "cmTC_2027d/fast"
> 
>   /usr/bin/make -f CMakeFiles/cmTC_2027d.dir/build.make
>   CMakeFiles/cmTC_2027d.dir/build
> 
>   make[1]: Entering directory
>   `/home/nntchau/aws-iot-device-sdk-cpp-1.1.1/build_share/CMakeFiles/CMakeTmp'
> 
> 
>   Building CXX object CMakeFiles/cmTC_2027d.dir/testCXXCompiler.cxx.o
> 
>   
>   //home/nntchau/Android/Sdk/ndk-bundle/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/bin/arm-linux-androideabi-g++
>   --target=armv7-none-linux-androideabi
>   --sysroot=//home/nntchau/Android/Sdk/ndk-bundle/platforms/android-9/arch-arm
>   -isystem
>   //home/nntchau/Android/Sdk/ndk-bundle/sources/cxx-stl/gnu-libstdc++/4.9/include
>   -isystem
>   //home/nntchau/Android/Sdk/ndk-bundle/sources/cxx-stl/gnu-libstdc++/4.9/libs/armeabi-v7a/include
>   -isystem
>   //home/nntchau/Android/Sdk/ndk-bundle/sources/cxx-stl/gnu-libstdc++/4.9/include/backward
>   -isystem
>   //home/nntchau/Android/Sdk/ndk-bundle/platforms/android-9/arch-arm/usr/include
>   -g -DANDROID -ffunction-sections -funwind-tables -fstack-protector-strong
>   -no-canonical-prefixes -march=armv7-a -mfloat-abi=softfp -mfpu=vfpv3-d16
>   -fno-integrated-as -mthumb -Wa,--noexecstack -Wformat
>   -Werror=format-security -O0 -fno-limit-debug-info -o
>   CMakeFiles/cmTC_2027d.dir/testCXXCompiler.cxx.o -c
>   /home/nntchau/aws-iot-device-sdk-cpp-1.1.1/build_share/CMakeFiles/CMakeTmp/testCXXCompiler.cxx
> 
> 
>   arm-linux-androideabi-g++: error: unrecognized command line option
>   '-fno-integrated-as'
> 
>   arm-linux-androideabi-g++: error: unrecognized command line option
>   '-fno-limit-debug-info'
> 
>   make[1]: *** [CMakeFiles/cmTC_2027d.dir/testCXXCompiler.cxx.o] Error 1
> 
>   make[1]: Leaving directory
>   `/home/nntchau/aws-iot-device-sdk-cpp-1.1.1/build_share/CMakeFiles/CMakeTmp'
> 
> 
>   make: *** [cmTC_2027d/fast] Error 2
> 
>   
> 
>   
> 
>   CMake will not be able to correctly generate this project.
> Call Stack (most recent call first):
>   CMakeLists.txt:2 (project)
>

**Comment by vareddy** [comment]
Hi @chaunnt,
Answers to your questions below
1. We are currently working on adding an example toolchain file to show how you can cross-compile it for any toolchain. It will be included in the SDK in a future release but we can't commit on the timeframe.
2. The output cmake that you pasted shows that the Android g++ compiler is broken. It fails as it does not recognize the flags. I'm not completely familiar with those flags or the Android toolchain, hence I need to do more research as to why it failed. Also, are you crosscompiling it using [Android Studio](https://developer.android.com/studio/projects/add-native-code.html) or on command line? 
3. We have not tested cross-compiling this library with Android hence I am not sure why it does not find the library. If you are looking at integrating AWS IoT with your Android app, have you considered using the AWS IoT Device SDK for Android? It is written natively in Java and will not require cross compilation. [Github link](https://github.com/aws/aws-sdk-android/tree/master/aws-android-sdk-iot) and the read me is [here](https://github.com/aws/aws-sdk-android/blob/master/README.md).
4. The library can be compiled as a shared object (.so) file by changing [this](https://github.com/aws/aws-iot-device-sdk-cpp/blob/master/CMakeLists.txt#L3) line to YES. Again, it depends on whether the target platform supports dynamic linking or else you will get the following warning
`CMake Warning (dev) at CMakeLists.txt:49 (add_library):
       ADD_LIBRARY called with SHARED option but the target platform does not support dynamic            linking. Building a STATIC library instead. This may lead to problems`

I know a lot of these are not definitive answers but these are specific to Android so I will have to investigate more into it. 
Please let us know if you have any more questions.
Thanks for using AWS IoT!
Varun

**Comment by chaunnt** [comment]
Hi @vareddy 

Thank you for your support.

> 1. We are currently working on adding an example toolchain file to show how you can cross-compile it for any toolchain. It will be included in the SDK in a future release but we can't commit on the timeframe.

Thank you. At least, we know it is possible.

> 2. The output cmake that you pasted shows that the Android g++ compiler is broken. It fails as it does not recognize the flags. I'm not completely familiar with those flags or the Android toolchain, hence I need to do more research as to why it failed. Also, are you crosscompiling it using Android Studio or on command line?

I'm using command line in Ubuntu 14.04

> 3. We have not tested cross-compiling this library with Android hence I am not sure why it does not find the library. If you are looking at integrating AWS IoT with your Android app, have you considered using the AWS IoT Device SDK for Android? It is written natively in Java and will not require cross compilation. Github link and the read me is here.

We are investigating this too. Sorry but I can not use aws-iot-sdk for Android. I'm using this library to implement another C++ Library which will be used in Android native app :) :) 

>4.  The library can be compiled as a shared object (.so) file by changing this line to YES. Again, it depends on whether the target platform supports dynamic linking or else you will get the following warning
> CMake Warning (dev) at CMakeLists.txt:49 (add_library): ADD_LIBRARY called with SHARED option but the target platform does not support dynamic linking. Building a STATIC library instead. This may lead to problems

Yes. We see that. we now can build & use shared library (.so) file. The remain problem is only cross-compiling

**Comment by vareddy** [comment]
Hi @chaunnt,
I did some testing with our toolchain file, using the Android toolchain, GNU version 4.9.0. The cmake is successful but I see other errors while running `make`, as shown below:

` error: 'to_string' is not a member of 'std'+ std::to_string(std::chrono::high_resolution_clock::now().time_since_epoch().count());`

The alternative might be to implement it by yourself as outlined in this StackOverflow post: https://stackoverflow.com/questions/22774009/android-ndk-stdto-string-support 

I have attached the toolchain file that I used for testing with this answer. Replace the path to the toolchain with where your toolchain is located. I believe everything else will stay the same.  Rename it to `.cmake` as that is what CMake expects. 

You set it up for cross compilation by running cmake as follows:
`cmake ../. -DCMAKE_TOOLCHAIN_FILE=../AndroidToolchain.cmake`
and run make as usual.

[AndroidToolchain.txt](https://github.com/aws/aws-iot-device-sdk-cpp/files/1206564/AndroidToolchain.txt)

If you are using OpenSSL, you need to replace the paths in the `network/CMakeLists.txt.in` with a cross-compiled version of OpenSSL for Android. 

Please let us know if this works and if you have any more questions.
Thanks!
Varun

**Comment by chaunnt** [comment]
Hi @vareddy 
Thank you. I do as same as you talked.

> If you are using OpenSSL, you need to replace the paths in the network/CMakeLists.txt.in with a cross-compiled version of OpenSSL for Android.

In CMakeLists.txt.in, I modify as below but error about OPENSSL still occur

	# Check for required packages (OpenSSL)
	#if using Mac OS, comment find_package(OpenSSL REQUIRED) and uncomment following two lines. Change the path to your openssl 1.0.2 directory.
	find_package(OpenSSL REQUIRED)
    set(OPENSSL_INCLUDE_DIR "/usr/local/openssl/include")
    set(OPENSSL_LIBRARIES "/usr/local/openssl/lib/libssl.a;/usr/local/openssl/lib/libcrypto.a")

I'm doing other way as below:
1. Run make-standalone-toolchain.sh in Android SDK -> Extract a toolchain folder (included sysroot)
2. Use cmake with this toolchain

> /opt/cmake/bin/cmake ../. -DCMAKE_SYSTEM_NAME=Android -DCMAKE_ANDROID_STANDALONE_TOOLCHAIN=/home/nntchau/Downloads/arm-linux-androideabi -DCMAKE_C_FLAGS="-std=c99"

But there is errors:
- UINT16_MAX is not defined -> I define it.

After that, there is another:

> /home/nntchau/aws-iot-device-sdk-cpp-1.1.1/src/discovery/Discovery.cpp:174:42: error: 
>       no member named 'stoul' in namespace 'std'
>             size_t content_length = std::stoul(content_length_str);

**Comment by vareddy** [comment]
Hi @chaunnt,

For the OpenSSL, you need to cross-compile it for Android and use that library. You cannot point it to the system's OpenSSL as that will not work for Android. See [here](https://wiki.openssl.org/index.php/Android) for more information on how to cross-compile OpenSSL. We depend on v1.0.2 of OpenSSL so that should be used in the cross-compile step. 

You need to  comment out this [line](https://github.com/aws/aws-iot-device-sdk-cpp/blob/master/network/CMakeLists.txt.in#L59) by adding a `#` before it and uncomment the two lines following it. You will then need to modify the paths to the  `libcrypto.a` and `libssl.a` to point to the cross-compiled OpenSSL for Android. 

Regarding the `stoul` issue, it looks like it could be a deep rooted problem in Android. Could you check if any of the solutions given in this [StackOverflow post](https://stackoverflow.com/questions/17950814/how-to-use-stdstoul-and-stdstoull-in-android) work?

Thanks!
Varun

**Comment by chaunnt** [comment]
Thank you @vareddy 

> Regarding the stoul issue, it looks like it could be a deep rooted problem in Android. Could you check if any of the solutions given in this StackOverflow post work?

I'm finding solution.

Did you finish cross-compile for Android ?

**Comment by vareddy** [comment]
Hi @chaunnt,
I haven't tried compiling the application completely, I have just tested the whether the CMake can find the compiler using the toolchain file. 
Did the toolchain file work for you and were you able to compile the application completely?
Thanks

**Comment by chaunnt** [comment]
Hi @vareddy 

> I haven't tried compiling the application completely, I have just tested the whether the CMake can find the compiler using the toolchain file.
> Did the toolchain file work for you and were you able to compile the application completely?

I'm still trying.
toolchain seem to have some error.
NDK of Android is lacking of support some function of C++ (std functions).
I tried work around from stackoverflow but no luck yet :) 

As they mentioned in forum, I will try to use crystax-ndk.

Note: because I'm very new to both Android NDK & Linux build (CMake as well). 
Then process may be a little slow.

**Comment by vareddy** [comment]
Hi @chaunnt ,
Were you able to make any progress with the cross compilation?
Varun

**Comment by vareddy** [comment]
Hi @chaunnt,
Since we haven't heard from you in a long time, I'm closing this issue. Please open [new issue](https://github.com/aws/aws-iot-device-sdk-cpp/issues/new) if you have any more questions about the C++ device SDK.
Thanks!
Varun

**Comment by aasick** [comment]
> Hi @chaunnt,
> Answers to your questions below
> 
> 1. We are currently working on adding an example toolchain file to show how you can cross-compile it for any toolchain. It will be included in the SDK in a future release but we can't commit on the timeframe.
> 2. The output cmake that you pasted shows that the Android g++ compiler is broken. It fails as it does not recognize the flags. I'm not completely familiar with those flags or the Android toolchain, hence I need to do more research as to why it failed. Also, are you crosscompiling it using [Android Studio](https://developer.android.com/studio/projects/add-native-code.html) or on command line?
> 3. We have not tested cross-compiling this library with Android hence I am not sure why it does not find the library. If you are looking at integrating AWS IoT with your Android app, have you considered using the AWS IoT Device SDK for Android? It is written natively in Java and will not require cross compilation. [Github link](https://github.com/aws/aws-sdk-android/tree/master/aws-android-sdk-iot) and the read me is [here](https://github.com/aws/aws-sdk-android/blob/master/README.md).
> 4. The library can be compiled as a shared object (.so) file by changing [this](https://github.com/aws/aws-iot-device-sdk-cpp/blob/master/CMakeLists.txt#L3) line to YES. Again, it depends on whether the target platform supports dynamic linking or else you will get the following warning
>    `CMake Warning (dev) at CMakeLists.txt:49 (add_library): ADD_LIBRARY called with SHARED option but the target platform does not support dynamic            linking. Building a STATIC library instead. This may lead to problems`
> 
> I know a lot of these are not definitive answers but these are specific to Android so I will have to investigate more into it.
> Please let us know if you have any more questions.
> Thanks for using AWS IoT!
> Varun

can u explain me how to cross compile for arm cortex A7 processor, with this "arm-linux-gnueabihf-g++" and what are the modification needed in cmake files
